contracts/proxy/ERC1967/ERC1967Upgrade.sol

```diff
@@ -82,28 +83,14 @@ abstract contract ERC1967Upgrade {
         bytes memory data,
         bool forceCall
     ) internal {
-        address oldImplementation = _getImplementation();
-
-        // Initial upgrade and setup call
-        _setImplementation(newImplementation);
-        if (data.length > 0 || forceCall) {
-            Address.functionDelegateCall(newImplementation, data);
-        }
-
-        // Perform rollback test if not already in progress
-        StorageSlot.BooleanSlot storage rollbackTesting = StorageSlot.getBooleanSlot(_ROLLBACK_SLOT);
-        if (!rollbackTesting.value) {
-            // Trigger rollback using upgradeTo from the new implementation
-            rollbackTesting.value = true;
-            Address.functionDelegateCall(
-                newImplementation,
-                abi.encodeWithSignature("upgradeTo(address)", oldImplementation)
+        if (StorageSlot.getBooleanSlot(_ROLLBACK_SLOT).value) {
```

Comment:
I don't think we should remove in 5.0. Removing it will prevent UUPS contract deployed with the "old" logic from being upgraded to the new one.

Reply:
The internal functions in `ERC1967Upgrade` are not virtual. This is because it was designed as a library (which it is not for some reason I cannot exactly remember, maybe the "defining events" part. This means you cannot override them.

Also, `_upgradeToAndCallSecure` needs to have access to `_setImplementation`, which is private. This is why it was in `ERC1967Upgrade` and not in `UUPSUpgradeable` from the start.

I'm ok with renaming it, but I don't think we should move it

Reply:
>Removing it will prevent UUPS contract deployed with the "old" logic from being upgraded to the new one.

The thing is there are likely to be other things that will prevent it anyway (storage compatibility mainly). And if needed, it would still be possible to upgrade in two steps.

Reply:
> there are likely to be other things that will prevent it anyway

I really don't think that is the right approach. We currently include a UUPSUpgradeable contract, which relies on a rollback mechanism, and which doesn't implement the proxiableUUID function. This was released in 4.1, allow with the UUPS proxy ... and people ARE using it.

I don't think it is acceptable to break that pattern in 4.5 and say "we change the way security is done, and you just can't upgrade you instance that is 3 month old without doing it in multiple steps". 
I may be ok to change the security mechanism from rollback testing to an 1822-like approach, but I'm really not with breaking the upgrade process that UUPS user rely on.

Reply:
Ok, let's discuss this again when 5.0 comes around. I lean towards deprecating it so we can simplify the code and reduce attack surface, but I appreciate the pushback.